src/main/JayZ.ts

```diff
@@ -1,49 +1,123 @@
-import { EncryptedJayZItem, Encryptor } from "./encryptors"
+import { memzero, ready } from "libsodium-wrappers"
+import { DataKey, DataKeyProvider } from "./DataKeyProvider"
+import { Encryptor } from "./Encryptor"
+import { LibsodiumEncryptor } from "./LibsodiumEncryptor"
+import { EncryptedItemMetadataV1, EncryptedJayZItem, LegacyEncryptedItemMetadata, MetadataVersion } from "./types"
 
-export interface JayZProps {
-  encryptor: Encryptor
+export type JayZConfig = {
+  keyProvider: DataKeyProvider
+  encryptor?: Encryptor
+  maxUsesPerDataKey?: number
 }
 
-export interface EncryptItemProps<T, U extends keyof T> {
+export type EncryptItemRequest<T, K extends keyof T> = {
   item: T
-  fieldsToEncrypt: U[]
+  fieldsToEncrypt: K[]
 }
 
 export class JayZ {
-  private encryptor: Encryptor
+  public readonly ready = ready
+  private keyProvider: DataKeyProvider
+  private encryptor: Encryptor = new LibsodiumEncryptor()
+  private maxUsesPerDataKey: number
+  private currentDataKey?: Promise<DataKey>
+  private currentDataKeyUsesRemaining: number
 
-  constructor(props: JayZProps) {
-    this.encryptor = props.encryptor
+  constructor(config: JayZConfig) {
+    this.keyProvider = config.keyProvider
+    this.encryptor =
+      config.encryptor !== undefined
+        ? config.encryptor
+        : new LibsodiumEncryptor()
+    this.maxUsesPerDataKey = config.maxUsesPerDataKey || 1
+    this.currentDataKeyUsesRemaining = this.maxUsesPerDataKey
   }
 
-  async encryptItem<T, U extends keyof T>(itemToEncrypt: EncryptItemProps<T, U>): Promise<EncryptedJayZItem<T, U>> {
+  async encryptItem<T, K extends keyof T>(
+    itemToEncrypt: EncryptItemRequest<T, K>
+  ): Promise<EncryptedJayZItem<T, K>> {
     const { item, fieldsToEncrypt } = itemToEncrypt
-    const { encryptedItem } = await this.encryptor.encrypt({ item, fieldsToEncrypt })
-    return encryptedItem
+    const { dataKey, encryptedDataKey } = await this.getNextDataKey()
+    const { plaintextFields, nonce, encryptedFields } = this.encryptor.encrypt({
+      item,
+      fieldsToEncrypt,
+      dataKey
+    })
+
+    const __jayz__metadata: EncryptedItemMetadataV1 = {
+      encryptedDataKey,
+      nonce,
+      scheme: this.encryptor.scheme,
+      encryptedFields,
+      metadataVersion: MetadataVersion.V1
+    }
+
+    return {
+      ...plaintextFields,
+      __jayz__metadata
+    }
   }
 
-  async encryptItems<T, U extends keyof T>(
-    itemsToEncrypt: EncryptItemProps<T, U>[]
-  ): Promise<EncryptedJayZItem<T, U>[]> {
+  encryptItems<T, K extends keyof T>(
```

Comment:
The `async` tag was removed here likely because there's no need to use `await` in the body of this function.